modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/AbstractArtifactClassLoaderConfigurationAssembler.java

```diff
@@ -0,0 +1,166 @@
+        loadUrls(getProjectFolder(), getBundleDependencies(), artifactClassLoaderConfigurationBuilder);
+    dependenciesArtifactsUrls.forEach(artifactClassLoaderConfigurationBuilder::containing);
+
+    // TODO: check if it belongs to the deny-list to decide whether local packages should be populated
+    populateLocalPackages(packagerClassLoaderModel, artifactClassLoaderConfigurationBuilder);
+
+    artifactClassLoaderConfigurationBuilder.dependingOn(new HashSet<>(getBundleDependencies()));
+
+    return artifactClassLoaderConfigurationBuilder.build();
+  }
+
+  private List<URL> loadUrls(File artifactFile,
+                             List<BundleDependency> dependencies,
+                             ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder) {
+    final List<URL> dependenciesArtifactsUrls = new ArrayList<>();
+
+    // TODO: consider artifact patches for the case this is run within a Runtime
+
+    final URL artifactFileUrl = getUrl(artifactFile, artifactFile);
+    dependenciesArtifactsUrls.add(artifactFileUrl);
+
+    dependenciesArtifactsUrls.addAll(addArtifactSpecificClassLoaderConfiguration(artifactClassLoaderConfigurationBuilder));
+    dependenciesArtifactsUrls.addAll(addDependenciesToClasspathUrls(artifactFile, dependencies));
+
+    return dependenciesArtifactsUrls;
+  }
+
+  protected Collection<URL> addArtifactSpecificClassLoaderConfiguration(ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder) {
+    return Collections.emptyList();
+  }
+
+  private List<URL> addDependenciesToClasspathUrls(File artifactFile,
+                                                   List<BundleDependency> dependencies) {
+    final List<URL> dependenciesArtifactsUrls = new ArrayList<>();
+
+    dependencies.stream()
+        .filter(dependency -> !MULE_PLUGIN_CLASSIFIER.equals(dependency.getDescriptor().getClassifier().orElse(null)))
+        .filter(dependency -> dependency.getBundleUri() != null)
+        .filter(dependency -> !validateMuleRuntimeSharedLibrary(dependency.getDescriptor().getGroupId(),
+                                                                dependency.getDescriptor().getArtifactId(),
+                                                                artifactFile.getName()))
+        .forEach(dependency -> {
+          final URL dependencyArtifactUrl;
+          try {
+            dependencyArtifactUrl = dependency.getBundleUri().toURL();
+          } catch (MalformedURLException e) {
+            throw new MuleRuntimeException(e);
+          }
+          dependenciesArtifactsUrls.add(dependencyArtifactUrl);
+        });
+
+    return dependenciesArtifactsUrls;
+  }
+
+  private URL getUrl(File artifactFile, File file) {
+    try {
+      return file.toURI().toURL();
+    } catch (MalformedURLException e) {
+      throw new ArtifactActivationException(createStaticMessage(format("There was an exception obtaining the URL for the artifact [%s], file [%s]",
```

Comment:
ArtifactActivationException is the one we're using everywhere for any error. Maybe we should enrich it somehow? Like adding what sptep of the process the failure occurred in (generating descriptors, classloaders, extension models...). Don't need to do this on this PR, but create a ticket to track this.

Reply:
Created user story [W-11202067](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000xnkGBYAY/view) to track this.